ZIP 317: Add Ironwood-pool Action fees as a revision to ZIP 317 (#1314)
zcash/zips: zips/zip-0317.rst

  Original-Authors: Aditya Bharadwaj
  Credits: Madars Virza
           Francisco Gindre

>   Status: [Revision 0] Active, [Revision 1: NU6.3] Draft, [Revision 2] Draft

Reviewer: ```suggestion
  Status: [Revision 0] Active, [Revision 1: NU6.3] Proposed, [Revision 2] Draft
```
(subject to ZIP Editor consensus in the 2026-06-30 sync meeting)

>   Status: [Revision 0] Active, [Revision 1: NU6.3] Proposed, [Revision 2] Draft

Author: ```suggestion
  Status: [Revision 0] Active, [Revision 1: NU6.3] Draft, [Revision 2] Draft
```

============

> * The conventional fee formula should not favour or discriminate against any

Reviewer: ```suggestion
* The conventional fee formula should not favour or discriminate against
```
(can't suggest across a removed line)

>   of the Ironwood, Orchard, Sapling, or transparent protocols.

Reviewer: ```suggestion
  transactions involving any of the Ironwood, Orchard, Sapling, or transparent
  pools.
```

Let $\mathsf{floor}(x)$ be the largest integer $\leq x$. |br|
Let $\mathsf{ceiling}(x)$ be the smallest integer $\geq x$.

Revisions
---------

.. _`Revision 0`:

* Revision 0: The initial fee mechanism specified by this ZIP, in effect

>   prior to the Ironwood network upgrade.

Reviewer: ```suggestion
  prior to NU6.3.
```

.. _`Revision 1`:

* Revision 1: Adds a contribution for Ironwood-pool [#zip-0229]_ Actions, to

>   be enacted at the Ironwood network upgrade (NU6.3 [#zip-0258]_).

Reviewer: ```suggestion
* Revision 1: Adds a contribution for *Ironwood-pool* [#zip-0229]_ Actions, to
  be enacted at NU6.3 which introduces the *Ironwood pool* [#zip-0258]_ [#Ironwood-book]_.
```

:math:`\mathit{grace\_actions}`                :math:`2`     logical actions
:math:`\mathit{p2pkh\_standard\_input\_size}`  :math:`150`   bytes
:math:`\mathit{p2pkh\_standard\_output\_size}` :math:`34`    bytes
============================================== ============= ==============================================

As of `Revision 0`_, the following contributions are defined:

|br|

.. math::

   \begin{array}{lcl}
     \mathit{contribution}_{\,\mathsf{Transparent}} &=& \mathsf{max}\big(\mathsf{ceiling}\big(\frac{\mathit{tx\_in\_total\_size}}{\mathit{p2pkh\_standard\_input\_size}}\big),\,
                                                                         \mathsf{ceiling}\big(\frac{\mathit{tx\_out\_total\_size}}{\mathit{p2pkh\_standard\_output\_size}}\big)\big) \\
     \mathit{contribution}_{\,\mathsf{Sprout}}      &=& 2 \cdot \mathit{nJoinSplit} \\
     \mathit{contribution}_{\,\mathsf{Sapling}}     &=& \mathsf{max}(\mathit{nSpendsSapling},\, \mathit{nOutputsSapling}) \\

>      \mathit{contribution}_{\,\mathsf{Orchard}}     &=& \mathit{nA\kern-0.1em ctionsOrchard}

Reviewer 2: pre-existing typo, definition and other refs need updating too.

```suggestion
     \mathit{contribution}_{\,\mathsf{Orchard}}     &=& \mathit{nA\kern-0.1em nActionsOrchard}
```

Author: This would duplicate the `nA`

Reviewer: It's correct as written (if a little presentation-oriented), including the space after "em" which terminates the macro.

:math:`\mathit{nSpendsSapling}`               number the number of Sapling spends
:math:`\mathit{nOutputsSapling}`              number the number of Sapling outputs
:math:`\mathit{nA\kern-0.1em ctionsOrchard}`  number the number of Orchard actions
============================================= ====== ====================================================================

As of `Revision 1`_, the $\mathit{logical\_actions}$ sum additionally includes
a contribution for Ironwood-pool [#zip-0229]_ Actions, which use the same
Action structure as Orchard-pool Actions:

|br|

.. math::

>    \mathit{contribution}_{\,\mathsf{Ironwood}} = \mathit{nA\kern-0.1em ctionsIronwood}

Reviewer 2: ```suggestion
   \mathit{contribution}_{\,\mathsf{Ironwood}} = \mathit{nA\kern-0.1em nActionsIronwood}
```

Author: This would duplicate the `nA`

The intention is to make the fee paid for a transaction depend on its

> impact on the network, without discriminating between different protocols

Reviewer: ```suggestion
The intention is to make the fee paid for a transaction depend on its impact
on the network, without discriminating between different transaction components
```

.. [#protocol-txnencoding] `Zcash Protocol Specification, Version 2022.3.8. Section 7.1: Transaction Encoding and Consensus <protocol/protocol.pdf#txnencoding>`_
.. [#sigop-limit] `zcash/zips issue #568 - Document block transparent sigops limit consensus rule <https://github.com/zcash/zips/issues/568>`_
.. [#madars-1] `Madars concrete soft-fork proposal <https://forum.zcashcommunity.com/t/zip-reduce-default-shielded-transaction-fee-to-1000-zats/37566/89>`_
.. [#zip-0229] `ZIP 229: Version 6 Transaction Format <zip-0229.md>`_
.. [#zip-0231] `ZIP 231: Memo Bundles <zip-0231.md>`_
.. [#zip-0248] `ZIP 248: Extensible Transaction Format (unmerged) <https://github.com/zcash/zips/pull/1156>`_
.. [#zip-0258] `ZIP 258: Deployment of the NU6.3 Network Upgrade <zip-0258.md>`_
.. [#zip-0313] `ZIP 313: Reduce Conventional Transaction Fee to 1000 zatoshis <zip-0313.rst>`_

> .. [#zip-0401] `ZIP 401: Addressing Mempool Denial-of-Service <zip-0401.rst>`_

Reviewer: ```suggestion
.. [#zip-0401] `ZIP 401: Addressing Mempool Denial-of-Service <zip-0401.rst>`_
.. [#Ironwood-book] `The Ironwood Book <https://zcash.github.io/ironwood/>`_
```
